src/main/java/org/jabref/gui/openoffice/OOBibBase.java

```diff
@@ -1,1422 +1,1691 @@
-        return getDocumentTitle(mxDoc);
+    /**
+     * Either return a valid DocumentConnection or throw
+     * NoDocumentException.
+     */
+    public DocumentConnection getDocumentConnectionOrThrow()
+        throws
+        NoDocumentException {
+        return this.connection.getDocumentConnectionOrThrow();
     }
 
-    private Optional<String> getDocumentTitle(XTextDocument doc) {
-        if (doc == null) {
-            return Optional.empty();
-        } else {
-            try {
-                return Optional.of(String.valueOf(OOUtil.getProperty(doc.getCurrentController().getFrame(), "Title")));
-            } catch (UnknownPropertyException | WrappedTargetException e) {
-                LOGGER.warn("Could not get document title", e);
-                return Optional.empty();
-            }
-        }
+    /**
+     *  The title of the current document, or Optional.empty()
+     */
+    public Optional<String> getCurrentDocumentTitle() {
+        return this.connection.getCurrentDocumentTitle();
     }
 
-    public void selectDocument() throws NoDocumentException, NoSuchElementException, WrappedTargetException {
-        List<XTextDocument> textDocumentList = getTextDocuments();
-        XTextDocument selected;
-        if (textDocumentList.isEmpty()) {
-            // No text documents found.
-            throw new NoDocumentException("No Writer documents found");
-        } else if (textDocumentList.size() == 1) {
-            // Get the only one
-            selected = textDocumentList.get(0);
-        } else {
-            // Bring up a dialog
-            selected = selectComponent(textDocumentList);
-        }
-
-        if (selected == null) {
-            return;
-        }
-        xCurrentComponent = UnoRuntime.queryInterface(XComponent.class, selected);
-        mxDoc = selected;
-
-        UnoRuntime.queryInterface(XDocumentIndexesSupplier.class, xCurrentComponent);
+    /* ****************************
+     *
+     *           Misc
+     *
+     * ****************************/
 
-        XModel xModel = UnoRuntime.queryInterface(XModel.class, xCurrentComponent);
-        XController xController = xModel.getCurrentController();
-        xViewCursorSupplier = UnoRuntime.queryInterface(XTextViewCursorSupplier.class, xController);
+    public List<CitationEntry> getCitationEntries()
+        throws
+        UnknownPropertyException,
+        WrappedTargetException,
+        NoDocumentException,
+        CreationException,
+        PropertyVetoException,
+        JabRefException {
 
-        // get a reference to the body text of the document
-        text = mxDoc.getText();
+        DocumentConnection documentConnection = this.getDocumentConnectionOrThrow();
 
-        // Access the text document's multi service factory:
-        mxDocFactory = UnoRuntime.queryInterface(XMultiServiceFactory.class, mxDoc);
+        checkIfOpenOfficeIsRecordingChanges(documentConnection);
```

Comment:
These (and some others) are testing preconditions for the GUI action they are called in.
I need to communicate to the GUI, or consider OOBibBase itself part of the GUI and
bring up a dialog. Most of OOBibBase did not do such things (except the select which document to connect part).

So the optiions I see are:

1) Move the precondition tests to the caller in (most likely) OOPanel.
Question: should the caller know about such details of the action?
Some of them could be truned into a warning:  for example
missing character and paragraph styles in the document:
we could still run, just skip applying those missing.
Or install a version and inform the user we did and that he can
edit them. 

2) Bring up dialog within the action.
Question: do error dialogs belong in the action implementation?

3) extend the return types of the actions to include diagnostics.
Something like this is already there in exportCitedHelper,
where the list of unresolvedKeys and the result database are packed together.
For the "cannot do it" cases this does not seem practical.

Reply:
>  1. It depends on the exception and the caller. 
> At the moment I am looking at `ManageCitationsDialogViewModel` 
> and I believe it is appropriate that it is aware of the failure to fetch 
> citation entries (e.g., use an optional
> `public Optional<List<CitationEntry>> getCitationEntries()`, 
> perhaps close the manage citation dialog if the optional is empty?).

Some of the checks are already in OpenOfficePanel, called before creating ManageCitationsDialogView, in
[manageCitations.setOnAction](https://github.com/antalk2/jabref/blob/57a1a973ad1dc38863431764a1019e0351461107/src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java#L283)

A problem (I think) with the code there
is that the constructor tries to get the data, but in some cases it cannot.

A constructor can only throw or return an instance encoding "I am not usable".

Technically it is not the constructor of ManageCitationsDialogView, it is its initialize() method,
which passes ooBase  to `new ManageCitationsDialogViewModel(ooBase, dialogService);`
which calls `ooBase.getCitationEntries();`

So deep down in dialog construction we start to ask for details
we sometimes cannot get.

I would suggest that we should not try to build and show a dialog we cannot fill.
I do not know how to close a dialog we are currently building and is (not itself, its parent) going to be
the argument (or was, depending on when initialize is called)
of showCustomDialogAndWait when we return from the constructor to initialize() and from there ... I do not know,
the caller of initialize(). 
The tests themselves may be duplicated to have a throwing and a returning version.

Ideas

(1): move the call `ooBase.getCitationEntries();` to before constructing
ManageCitationsDialogView. We can then decide based on its return value
what todo. If it succeeds, we can pass the data to the dialog constructor (we still need to pass ooBase,
as long as the dialog wants to call `ooBase.applyCitationEntries`)

(2) add `ooBase.getCitationEntriesPretest();` to be called before the constructor.
But in this case `ooBase.getCitationEntries();` might still fail for reasons we did not or could not
test without actually tyring to collect the data.


> I didn't notice that OOBibBase was part of the GUI (`gui/openoffice`),
> but since it is I am assuming it is intended as a ViewModel
> and I'd consider it appropriate that it is the one that shows an error dialog.

I think it probaly predates the gui/logic separation, or as a plugin it was not considered
subject to it. A lot of its content does not belong to gui, but was not easy to separate, so it stayed
in gui due to its use of dialog. 

>  2. What is "the action"? The lambda functions associated with a button?
> If that is what you referred to, I'd say preliminarily yes.
> If there are larger chunks of code you can break them out in their own method as well,
> similar to `OpenOfficePanel#connectAutomatically`.

You pointed to a method in OOBibBase, which is technically in gui, but mostly consists
of methods that should be in logic. In my head there is a separation between "real gui"
that is about opening dialogs, and methods that "implement the actions" (I referred to these).
These need to be careful and test what they have (connection, necessary conditions to work)
 in order to avoid messing up the document by failing during modifications. It feels safer to include the checks inside,
otherwise we rely on the gui calling the necessary checks (while the list of those may still evolve). 
But what I thought of as "action implementation" can still be split to "actual implementation" and a wrapper
that does the checks and shows dialogs. Maybe OpenOfficePanel could export 
showNoDocumentErrorMessage() and other methods (or they could be moved out)
and delegate calling them to other modules.

It is probably not the right time to start reorganizing even more files.

What do you think about the `(1): move the call `ooBase.getCitationEntries();` to before constructing`
solution?

> > extend the return types of the actions to include diagnostics.
>  3. There are some cases where this is appropriate. 
> As a personal preference, I try to avoid it, but it isn't always possible without making the code unreadable.

In this respect java is strange: there is Optional with map, but this latter
will not allow checked exceptions. It does not have Pair (except where they needed it) and it does not have
Either<T,U> or Result<Value,Messages> which could allow such extension wihout
creating a new class for each case.

Reply:
(Part of) The reason why I am harping about returning an Optional, is that using JavaFX we can [bind to an optional value](https://github.com/tobiasdiez/EasyBind#optional-observable-values) in the ViewModel and display an error message and call `this.close()` if it ever becomes `Optional.empty()` while the Window/Dialog is shown. Say something happens while you have the window/dialog open (e.g., in a background thread, if you manually refresh the list), and we know you can no longer access the document. I am not too fond of JavaFX but it is awfully convenient sometimes.

>  In my head there is a separation between "real gui"
that is about opening dialogs, and methods that "implement the actions" (I referred to these).

You are referring to View vs ViewModel, where View is the "real GUI"?

> I do not know how to close a dialog we are currently building and is (not itself, its parent) going to be
> the argument (or was, depending on when initialize is called)
> of showCustomDialogAndWait when we return from the constructor to initialize() and from there ... I do not know,
> the caller of initialize().

Can't we call this.close() in the constructor? ManageCitationDialogView has access to `viewModel.citationsProperty()` (which can be made Optional).

> What do you think about the (1): move the call ooBase.getCitationEntries(); to before constructing
solution?

I think I'll have to get back to you on that one. Preliminarily I'd consider OpenOfficePanel to be part of the View even if it seems to contain more logic than it should.

You are right in that perhaps now isn't the time for more refactoring. I'll have to think about it it, sometimes small refactoring of old code leads to a lot fewer total changes. I'll start digging into in OpenOfficePanel (instead of ManageCitationDialogView) and get back to you.

Reply:
> Maybe OpenOfficePanel could export
showNoDocumentErrorMessage() and other methods (or they could be moved out)
and delegate calling them to other modules.

I'd guess the "or they could be moved out" will be the easiest way of avoiding duplicating code.

Reply:
> (Part of) The reason why I am harping about returning an Optional, is that using JavaFX we can [bind to an optional value](https://github.com/tobiasdiez/EasyBind#optional-observable-values) in the ViewModel

You mean an invisible "property" and an addListener?

> and display an error message and call `this.close()` if it ever becomes `Optional.empty()` while the Window/Dialog is shown.

Does this work in initialize? Or only after we returned from it (or from the constructor?)?

> Say something happens while you have the window/dialog open (e.g., in a background thread
> or if you manually refresh the list), and we know you can no longer access the document.
> I am not too fond of JavaFX but it is awfully convenient sometimes.

To manually refresh the list only seems meaningful if we pushed some changes to the document.
We do not do that now, the only action that goes back to the document also closes the window.

> 
> > In my head there is a separation between "real gui"
> > that is about opening dialogs, and methods that "implement the actions" (I referred to these).
> 
> You are referring to View vs ViewModel, where View is the "real GUI"?

No, I think of the ViewModel as part of the "real gui". We leave that when it calls
to methods that work on the document and are not using dialogs or GUI properties.
I this context the separation seems to be:
GUI provides the style, preference values, list of databases and reference to ooBase,
and calls methods that only use these and the document. 

> > I do not know how to close a dialog we are currently building and is (not itself, its parent) going to be
> > the argument (or was, depending on when initialize is called)
> > of showCustomDialogAndWait when we return from the constructor to initialize() and from there ... I do not know,
> > the caller of initialize().
> 
> Can't we call this.close() in the constructor?

Is this a question, or a statement in question form?
I do not know: if at the moment we call close() the window
is not yet set up properly, the result might be unexpected.
But if close() just emits an event, then it might be in queue
until we return and the window is set up. Closing a window
within its constructor is not necessarily what the library authors expected
to happen.

> ManageCitationDialogView has access to `viewModel.citationsProperty()` (which can be made Optional).

I guess then initialize could add the listener after it constructed the viewModel.
That would cover the later updates. If we got Optional.empty() within the viewModel
constructor, the listener will probably not called. With a direct check we are back to
calling close from initialze()

> > What do you think about the (1): move the call ooBase.getCitationEntries(); to before constructing
> > solution?
> 
> I think I'll have to get back to you on that one.
> Preliminarily I'd consider OpenOfficePanel to be part of the View
> even if it seems to contain more logic than it should (which is why I am hesitating about adding more checks there).

We have different views of the dialog. Your image seems to be that things happen while it is open.
Probably apropriate for the case when multiple things can be done without closing the dialog.
My view is more of a temporary dialog: we send a list in and get an updated list back.
The dialog could return this list when closed and OpenOfficePanel could call OOBibBase with the result.
This, of course could not work if the dialog allowed several different actions modifying and rereading the
document. Is this the direction we expect this dialog to move?
 
> You are right in that perhaps now isn't the time for more refactoring. I'll have to think about it it, sometimes small refactoring of old code leads to a lot fewer total changes. I'll start digging into in OpenOfficePanel (instead of ManageCitationDialogView) and get back to you.

I started to move catching the exceptions (and partly not throwing them) to OOBiBBase. 
The caller becomes cleaner this way (it has nothing to catch), while the part in OOBibBase becomes more complicated.
It seems meaningful to move handling closer to those throwing.

Reply:
**Experiment with `ManageCitationsDialogView.close()`**

> Can't we call this.close() in the constructor?

We can call it from initialize()

Of the attempts [here](https://github.com/antalk2/jabref/blob/84020c93067e8647cd8227167cb1200cd7de919e/src/main/java/org/jabref/gui/openoffice/ManageCitationsDialogView.java#L66) only the last (variant 4, using Platform.runLater())
succeeded in closing the window, although it did not prevent it from appearing for a fraction of a second.
The others did not have an observable effect.
(The `ManageCitationsDialogView thisDialog = this;` line is only needed once.)